### ISSUE TITLE
perf: GAN tape-mode training is 5-10× slower than PyTorch CPU baseline — DCGAN MoreData_ShouldNotDegrade times out at 250 iters

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1390]
## Summary

`DCGANTests.MoreData_ShouldNotDegrade` (50 short + 200 long = 250 adversarial-train iterations on a 64×64 RGB DCGAN) cannot complete in the 120 s `[Fact(Timeout=120000)]` budget. Substep profiling shows AiDotNet is ~5-10× slower than PyTorch CPU at this scale, and the gap is concentrated in tape-mode forward + backward through the deep deconv/conv chain.

Surfaced while working PR #1290 CI Cluster 1 ([issue #1309](https://github.com/ooples/AiDotNet/issues/1309)) — the validator and conv-deserialize root-cause bugs in that cluster are fixed by [PR #1389](https://github.com/ooples/AiDotNet/pull/1389), but `MoreData_ShouldNotDegrade` still fails for a *perf* reason that's orthogonal to cluster-1's shape-contract scope.

## Substep profile (Release, net10.0)

DCGAN<double>: latent=100, image=3x64x64, gen feature maps=64, disc feature maps=64. Single Train step on synthetic random data, mean of 10 steps after 2-step warmup:

| Substep | Time | % |
|---|---:|---:|
| Generator.Predict (no-grad forward) | 19.3 ms | 3.7% |
| Discriminator.Train (forward + backward + optim, batch=2) | 186.5 ms | 35.9% |
| Generator adversarial (tape gen+disc forward + backward + optim) | 313.3 ms | 60.4% |
| **Total** | **519 ms / step** | |
| **Projected 250 iters** | **130 s** | (10 s over 120 s budget) |

For comparison: PyTorch DCGAN CPU at this scale is typically 30-100 ms/step total.

## What's likely contributing

1. **Generator forward runs twice per step**: once detached (`Generator.Predict`, NoGradScope) for the discriminator-training step's fake batch, then again on the tape (`Generator.TrainWithCustomLoss`) for the adversarial backward. PyTorch fuses these via `fake.detach()` — one forward, two consumers. Saves ~19 ms / step (4% if it's the only fix).
2. **Tape-mode backward through Conv2D**: AiDotNet uses BLAS-backed im2col+GEMM (`Conv2DWithIm2ColDouble` / `SimdGemm.Sgemm`). Suspect the **per-op tape recording** overhead — each Engine op records a GradFn closure, allocates GradNode state, walks `DifferentiableOps.RecordBinary` — adds material time per layer per direction.
3. **Discriminator.Train discriminator chain**: 187 ms for batch=2 through 11 layers ≈ 17 ms/layer/direction. PyTorch on similar shapes is < 1 ms/layer/direction on CPU. Even allowing 4× for the in-tree GEMM vs Intel MKL, there's ~3-4× left to explain.
4. **Double precision (`T=double`) penalty**: model-family tests use `INeuralNetworkModel<double>` so SIMD lanes per Vector256 are 4 instead of 8 (FP32). At paper-scale FP32 this would likely be sub-150 ms/step.

## Reproducer (timing probe)

The exact harness used to gather the profile lived in `tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/DCGANTimingProbe.cs` (kept out of PR #1389; reproducible by adding a Stopwatch around each substep in `GenerativeAdversarialNetwork.Train`).

```
=== measure (10 steps) ===
step  1:          471.5 ms
step  2:          570.7 ms
step  3:          483.3 ms
step  4:          484.4 ms
step  5:          511.3 ms
step  6:          512.3 ms
step  7:          496.7 ms
step  8:          540.4 ms
step  9:          544.1 ms
step 10:          576.4 ms
mean / step:     519.1 ms
projected 250 iters:  129.8 s (120 s timeout budget)
```

## Expected scope

- **Phase 1**: detach + share `fakeImages` between disc-step and gen-adversarial step (eliminates one full generator forward). Estimated win: 4%.
- **Phase 2**: profile the per-layer tape-record + GradNode allocation cost. Likely candidates: `DifferentiableOps.RecordBinary` / `RecordUnary` allocator pressure, GradNode pooling. Estimated win: 10-30%.
- **Phase 3**: Tensors-side Conv2DBackward kernel tuning at small N (the 4×4 deconv stride=2 padding=1 shape is non-canonical for the existing fast paths). Estimated win: 30-50%.

Closing the full gap likely requires all three phases.

## Affected tests

- `DCGANTests.MoreData_ShouldNotDegrade` (current visible victim)
- Any other GAN derivative that inherits `MoreDataShortIterations => 50` / `MoreDataLongIterations => 200` from `NeuralNetworkModelTestBase` without an override. `ConditionalGANTests` has a 1/2/4 override that masks the gap on its smaller fixture.

## NOT this issue

The shape-contract failures in #1309 (DCGAN validator deferred-shape + conv-deserialize lazy-input-depth) are unrelated and fixed by PR #1389.

**Comment by ooples** [comment]
Filed a Tensors-repo plan that tracks the work for this perf gap: ooples/AiDotNet.Tensors#403.

After re-checking the file boundary, the actual hot paths during a DCGAN training step (Conv2D im2col + GEMM, `DifferentiableOps.RecordBinary`, `GradNode` allocation, backward closures, `CompiledBackwardWalk`, microkernels) all live in AiDotNet.Tensors. AiDotNet.GenerativeAdversarialNetwork.Train() is thin orchestration. So the Tensors issue covers Phases 2 and 3 (autograd allocation cleanup + Conv2D small-N kernel) plus two additions:
- Phase D: TrainingWeightRegistry (extends Sub-E's FrozenWeightRegistry to training; auto-MarkDirty after optimizer step). ~3-8%.
- Phase E: FP64 microkernel verification (Sub-E only touched FP32 scalar; need to confirm AVX2 FP64 microkernel actually fires for the conv-im2col shapes).

Phase 1 (detach + share fakeImages) remains on this issue — that's pure AiDotNet orchestration.

Sequencing recommendation in #403: investigation first (1-2 days, shape catalog + microbench + alloc trace), then the higher-ROI kernel work can be prioritized against the alloc cleanup based on data.
